std/experimental/safeint.d

```diff
@@ -0,0 +1,1164 @@
+            assert(lessEqual(to!T(1), to!S(1)));
+            assert(lessEqual(to!T(-1), to!S(1)));
+            assert(!lessEqual(to!T(1), to!S(0)));
+
+            assert(!greater(to!T(1), to!T(1)));
+            assert(greater(to!T(2), to!T(1)));
+            assert(!greater(to!S(1), to!S(1)));
+            assert(greater(to!S(2), to!S(1)));
+            assert(!greater(to!T(1), to!S(1)));
+            assert(!greater(to!T(-1), to!S(1)));
+
+            assert(greaterEqual(to!T(1), to!T(1)));
+            assert(greaterEqual(to!T(2), to!T(1)));
+            assert(!greaterEqual(to!S(1), to!S(2)));
+            assert(!greaterEqual(to!T(-1), to!S(1)));
+        }
+    }
+}
+
+private pure auto getValue(T)(T t)
+{
+    static if (isIntegral!T)
+        return t;
+    else
+        return t.store.value;
+}
+
+/* Checks if the value of $(D s) can be stored by a variable of
+type $(D T).
+
+Params:
+    s = the value to check
+
+Returns:
+    $(D true) if the value can be stored, false otherwise.
+*/
+bool canConvertTo(T, S)(in S s) nothrow @nogc pure
+        if (isIntegral!(Unqual!T) && isIntegral!(SafeIntType!S))
+{
+    return (less(getValue(s), T.min) || greater(getValue(s), T.max)) ? false : true;
+}
+
+///
+pure unittest
+{
+    assert(canConvertTo!int(1337));
+    assert(canConvertTo!uint(1337));
+    assert(canConvertTo!int(-1337));
+    assert(!canConvertTo!uint(-1337));
+    assert(!canConvertTo!byte(1337));
+    assert(!canConvertTo!ubyte(1337));
+    assert(!canConvertTo!byte(-1337));
+    assert(!canConvertTo!ubyte(-1337));
+}
+
+pure unittest
+{
+    foreach (T; TTest)
+    {
+        foreach (S; TTest)
+        {
+            auto s = SafeInt!S(0);
+            assert(!s.isNaN);
+            assert(canConvertTo!T(s));
+        }
+    }
+}
+
+/** Returns the integer type used by a $(D SafeInt) to store the
+value.
+
+If an integer type is passed this type will be returned.
+*/
+template SafeIntType(T)
+{
+    static if (isIntegral!T)
+        alias SafeIntType = T;
+    else
+        alias SafeIntType = typeof(T.store.value);
+}
+
+///
+pure unittest
+{
+    static assert(is(SafeIntType!(SafeInt!int) == int));
+    static assert(is(SafeIntType!int == int));
+}
+
+/** Checks if the passed type is a $(D SafeInt).
+
+Returns:
+    $(D true) if the passed type $(D T) is an $(D SafeInt), false
+    overwise.
+*/
+template isSafeInt(T)
+{
+    //pragma(msg, typeof(T));
+    static if (is(T : SafeIntImpl!(S), S...))
+        enum isSafeInt = true;
+    else
+        enum isSafeInt = false;
+}
+
+///
+pure unittest
+{
+    //static assert( isSafeInt!(SafeIntInline!int));
+    //static assert( isSafeInt!(SafeIntExplicit!int));
+    static assert( isSafeInt!(SafeInt!int));
+    static assert(!isSafeInt!int);
+}
+
+pure unittest
+{
+    static assert(!isSafeInt!int);
+    foreach (T; TTest)
+    {
+        alias ST = SafeInt!T;
+        static assert(isSafeInt!ST);
+    }
+}
+
+struct SafeIntInline(T) {
+    T value = nan;
+
```

Comment:
I believe you are missing an `alias value this;` in `SafeIntInline`. (My benchmark does not compile without it.)